### ISSUE TITLE
Optional properties

### ISSUE THREAD

**Issue by MattNguyen** [interagent/prmd#259]
I currently use STI for several of my models. Some subclasses have different properties from the parent class, but one API endpoint may return both.

When I attempt to specify required properties, PRMD complains about the optional property not being specified. For example,

Given
```json
{
  "properties": {
    "propertyB": {...},
    "propertyA": {...}
  },
  "required": ["propertyA"],
}
```

PRMD will raise the following error:
```
#/data/attributes: failed schema #/definitions/user/definitions/resource/properties/attributes: "propertyB" weren't supplied.
```

Any suggestions?

**Comment by geemus** [comment]
I haven't worked with STI in this context (yet), so I don't know off hand. I don't see anything obvious from the context provided here, but might have better luck with a more complete schema. Could you provide more of that so I can work backwards? Thanks!

**Comment by fdr** [comment]
I'm having this problem with two serializations of one object in Pliny (where one is more detailed/expensive than another). One is called "default" and the other "detailed". For example:

```ruby
class Serializers::Dog < Serializers::Base
  def self.base(arg)
    {
      created_at:   arg.created_at.try(:iso8601),
      id:           arg.id,
      updated_at:   arg.updated_at.try(:iso8601),
      license_number: arg.license_number
    }
  end

  structure(:default) do |arg|
    base(arg)
  end

  structure(:detailed) do |arg|
    ret = base(arg)
    ret[:puppies] =
      Serializers::Dog.new(:default).serialize(arg.puppies)
    ret
  end
end
```

In general, my use case is I want to run this kind of thing in an endpoint, wherein the index of dogs uses the more compact representation, and the single-subject endpoint expands useful information:

```ruby
    namespace "/dogs" do
      get do
        encode serialize(Dog.all)
      end

      get "/:id" do |id|
        dog = Dog.first(id: id) || halt(404)
        encode serialize(dog, :detailed)
      end
  end
```

**Comment by geemus** [comment]
Yeah, sorry for my delay in responding. We mostly have built with an assumption of one-serialization per resource, which obviously is not super helpful for you. The pattern you describe is not uncommon, so we should probably find a sensible way to accommodate this. In the mean time, the best bet might be to define it as two resources or you may be able to get near enough (at least in terms of human interpreted docs) by explicitly specifying a different example on the detailed action.

**Comment by fdr** [comment]
I think that's too much code. Also, semantic bloat (DogWithPuppies vs. Dog?)

Instead, what I've done is make that field null in the summary serialization and push the complexity of knowing that to the reader.

**Comment by geemus** [comment]
Agreed, it isn't great but it was the first workaround that came to mind. I think we have also used nullables in some cases, though not for this case in particular. My supposition had been that not varying the keys in serializations had some value in terms of consistency/etc. But that implies that the more full serialization is not significantly more costly to produce (not always the case) and/or not way larger over the wire (where what that constitutes is notably ambiguous).
